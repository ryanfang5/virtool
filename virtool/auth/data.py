from typing import List

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession

from virtool.auth.models import SQLPermission
from virtool_core.models.auth import PermissionMinimal


class AuthData:
    def __init__(self, pg: AsyncEngine):
        self._pg = pg

    async def find(self, resource_type) -> List[PermissionMinimal]:
        """
        List all possible permissions.

        :return: a list of all permissions

        """
        statement = select(SQLPermission)

        if resource_type:
            statement = statement.filter_by(resource_type=resource_type)

        async with AsyncSession(self._pg) as session:
            result = await session.execute(statement)

        return [
            PermissionMinimal(**permission.to_dict()) for permission in result.scalars()
        ]
