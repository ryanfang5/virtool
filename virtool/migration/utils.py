from pathlib import Path


def get_revisions_path() -> Path:
    """
    Get the hardcoded path to data revision files.

    :return: the revisions path

    """
    return Path(__file__).parent.parent.parent / "assets/revisions"
