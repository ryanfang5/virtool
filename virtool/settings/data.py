
from virtool_core.models.settings import Settings

from virtool.mongo.core import DB
from virtool.settings.db import get, update
from virtool.settings.oas import UpdateSettingsSchema


class SettingsData:
    def __init__(self, db):
        self._db: DB = db

    async def get_all(self) -> Settings:
        """
        List all settings.

        :return: settings object

        """

        settings = Settings(**await get(self._db))

        return settings

    async def update(self, data: UpdateSettingsSchema) -> Settings:
        """
        Update the settings.

        :param data: updates to the current settings
        :return: the settings
        """

        data = data.dict(exclude_unset=True)

        settings = await update(self._db, data)

        settings.pop("software_channel", None)

        settings = Settings(**settings)

        return settings
