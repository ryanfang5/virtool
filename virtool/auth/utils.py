import sys
from enum import Enum
from logging import getLogger

import openfga_sdk
from aiohttp import ClientConnectorError
from openfga_sdk import (
    CreateStoreRequest,
    WriteAuthorizationModelRequest,
    TypeDefinition,
    Userset,
    OpenFgaApi,
    WriteRequest,
    TupleKey,
    ApiException,
    TupleKeys,
)
from openfga_sdk.api import open_fga_api

from virtool.auth.permissions import (
    AppPermission,
)

logger = getLogger("openfga")


async def connect_openfga(openfga_host: str, openfga_scheme: str):
    """
    Connects to an OpenFGA server and configures the store id.
    Returns the application client instance.
    """
    configuration = openfga_sdk.Configuration(
        api_scheme=openfga_scheme, api_host=openfga_host
    )

    logger.info("Connecting to OpenFGA")

    try:
        api_client = openfga_sdk.ApiClient(configuration)

        api_instance = open_fga_api.OpenFgaApi(api_client)

        configuration.store_id = await get_or_create_store(api_instance)

        await write_auth_model(api_instance)

    except ClientConnectorError:
        logger.critical("Could not connect")
        sys.exit(1)

    return api_instance


async def get_or_create_store(api_instance: OpenFgaApi):
    """
    Get the OpenFGA Store or create one if it does not exist.

    :return: the store id
    """
    response = await api_instance.list_stores()

    logger.info("Connected to OpenFGA")

    if response.stores:
        logger.info("Found existing store")
        return response.stores[0].id

    body = CreateStoreRequest(
        name="Virtool",
    )
    logger.info("Creating store")
    response = await api_instance.create_store(body)
    return response.id


async def write_auth_model(api_instance: OpenFgaApi):
    """
    Write the authorization model for the OpenFGA Store if it does not exist.
    """
    response = await api_instance.read_authorization_models()

    app_definition = TypeDefinition(
        type="app",
        relations={permission.name: Userset(this={}) for permission in AppPermission},
    )

    group_definition = TypeDefinition(
        type="group",
        relations={"member": Userset(this={})},
    )

    type_definitions = WriteAuthorizationModelRequest(
        type_definitions=[app_definition, group_definition],
    )

    if (
        response.authorization_models
        and response.authorization_models[0].type_definitions
        == type_definitions.type_definitions
    ):
        return

    await api_instance.write_authorization_model(type_definitions)


async def write_tuple(
    api_instance, user_type, user_id, relations, object_type, object_name
):
    """
    Write a relationship tuple in OpenFGA.
    """

    tuple_list = [
        TupleKey(
            user=f"{user_type}:{user_id}",
            relation=relation.name if isinstance(relation, Enum) else relation,
            object=f"{object_type}:{object_name}",
        )
        for relation in relations
    ]

    body = WriteRequest(writes=TupleKeys(tuple_keys=tuple_list))

    try:
        await api_instance.write(body)
    except ApiException:
        pass
