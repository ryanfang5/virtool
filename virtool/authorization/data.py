from typing import List

from virtool_core.models.auth import PermissionMinimal

from virtool.authorization.permissions import AppPermission, SpacePermission


class AuthorizationData:
    @staticmethod
    async def find() -> List[PermissionMinimal]:
        """
        List all possible permissions.

        :return: a list of all permissions

        """
        return [
            *[
                PermissionMinimal(
                    id=permission.name,
                    name=permission.value.name,
                    resource_type=permission.value.resource_type,
                    action=permission.value.action,
                    description=permission.value.description,
                )
                for permission in [
                    *[p for p in AppPermission],
                    *[p for p in SpacePermission],
                ]
            ]
        ]
