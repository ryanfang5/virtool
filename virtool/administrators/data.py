from typing import TYPE_CHECKING

from virtool_core.models.roles import AdministratorRole
from virtool_core.models.administrator import Administrator, AdministratorSearch, AdministratorMinimal
from virtool_core.utils import document_enum

from virtool.administrators.oas import (
    CreateAdministratorRequest,
    UpdateAdministratorRequest,
)
from virtool.authorization.client import AuthorizationClient
from virtool.authorization.relationships import AdministratorRoleAssignment
from virtool.data.errors import (
    ResourceNotFoundError,
)

if TYPE_CHECKING:
    from virtool.mongo.core import DB


AVAILABLE_ROLES = [
    {"id": role, "name": role.capitalize(), "description": role.__doc__}
    for role in document_enum(AdministratorRole)
]


class AdministratorsData:
    def __init__(self, authorization_client: AuthorizationClient, db: "DB"):
        self._authorization_client = authorization_client
        self._db = db

    async def find(self) -> AdministratorSearch:
        """
        List all administrators.

        :return: a list of all administrators

        """
        admin_list = await self._authorization_client.list_administrators()

        return AdministratorSearch(
            **{
                "items": [
                    AdministratorMinimal(
                        **{
                            **await self._db.users.find_one(admin_tuple[0]),
                            "role": AdministratorRole(admin_tuple[1]),
                        }
                    )
                    for admin_tuple in admin_list
                ],
                "available_roles": AVAILABLE_ROLES,
            }
        )

    async def get(self, user_id: str) -> Administrator:
        """
        Get an administrator.

        :param user_id: the user ID
        :return: the administrator
        """

        if admin_tuple := await self._db.users.find_one(
            user_id
        ) and await self._authorization_client.get_administrator(user_id):

            return Administrator(
                **{
                    **await self._db.users.find_one(user_id),
                    "role": AdministratorRole(admin_tuple[1]),
                    "available_roles": AVAILABLE_ROLES,
                }
            )

        raise ResourceNotFoundError()

    async def create(self, data: CreateAdministratorRequest) -> Administrator:
        """
        Add a user as an administrator with a given role.

        Set the user's administrator flag to true if given the full role.

        :param data: fields to add a user as an administrator
        :return: the administrator
        """

        if await self._db.users.find_one(data.user_id):

            if data.role == AdministratorRole.FULL:
                async with self._db.create_session() as session:
                    await self._db.users.update_one(
                        {"_id": data.user_id},
                        {"$set": {"administrator": True}},
                        session=session,
                    )

            await self._authorization_client.add(
                AdministratorRoleAssignment(data.user_id, AdministratorRole(data.role))
            )

            return Administrator(
                **{
                    **await self._db.users.find_one(data.user_id),
                    "role": AdministratorRole(data.role),
                    "available_roles": AVAILABLE_ROLES,
                }
            )

        raise ResourceNotFoundError()

    async def update(
        self, user_id: str, data: UpdateAdministratorRequest
    ) -> Administrator:
        """
        Update an administrator's role.

        Set the user's administrator flag to true if given the full role.

        Set the user's administrator flag to false if the full role is removed.

        :param user_id: the user ID of the administrator to update
        :param data: updates to the administrator's current role
        :return: the administrator
        """

        if await self._db.users.find_one(
            user_id
        ) and await self._authorization_client.get_administrator(user_id):

            async with self._db.create_session() as session:
                await self._db.users.update_one(
                    {"_id": user_id},
                    {
                        "$set": {
                            "administrator": data.role == AdministratorRole.FULL
                        }
                    },
                    session=session,
                )

            await self._authorization_client.add(
                AdministratorRoleAssignment(user_id, AdministratorRole(data.role))
            )

            return Administrator(
                **{
                    **await self._db.users.find_one(user_id),
                    "role": AdministratorRole(data.role),
                    "available_roles": AVAILABLE_ROLES,
                }
            )

        raise ResourceNotFoundError()

    async def delete(self, user_id: str):
        """
        Remove an administrator.

        :param user_id: the user ID
        """
        if admin_tuple := await self._db.users.find_one(
            user_id
        ) and await self._authorization_client.get_administrator(user_id):

            async with self._db.create_session() as session:
                await self._db.users.update_one(
                    {"_id": user_id},
                    {"$set": {"administrator": False}},
                    session=session,
                )

            await self._authorization_client.remove(
                AdministratorRoleAssignment(user_id, AdministratorRole(admin_tuple[1]))
            )

            return

        raise ResourceNotFoundError()
