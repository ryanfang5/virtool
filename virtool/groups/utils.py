from __future__ import annotations

from copy import deepcopy
from typing import List, Dict, Tuple

import virtool.users.utils
from virtool.authorization.permissions import SpacePermission
from virtool.authorization.relationships import GroupPermission
from virtool.users.utils import Permission


def merge_group_permissions(groups: List[dict]) -> Dict[str, bool]:
    """
    Return a :class:`dict` of permissions that will be
    inherited by a user belonging to all the passed ``groups``.

    :param groups: a list of group documents.
    :return: a dict keyed by permission names with boolean values
    indicating the state of the permission

    """
    permission_dict = virtool.users.utils.generate_base_permissions()

    for p in Permission:
        for group in groups:
            try:
                if group["permissions"][p.value]:
                    permission_dict[p.value] = True
            except KeyError:
                permission_dict[p.value] = False

    return permission_dict


def adapt_legacy_permissions_dict(permissions_dict: dict[str, bool]):
    adapted = deepcopy(permissions_dict)

    adapted.pop("remove_job", None)

    adapted["create_reference"] = adapted.pop("create_ref")
    adapted["delete_upload"] = adapted.pop("remove_file")
    adapted["create_upload"] = adapted.pop("upload_file")

    return adapted


def convert_permissions_dict_to_relationships(
    group_id: str,
    permissions_dict: dict[str, bool],
) -> Tuple[List[GroupPermission], List[GroupPermission]]:
    adapted_permissions_dict = adapt_legacy_permissions_dict(permissions_dict)

    adds = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if value
    ]

    removes = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if not value
    ]

    return adds, removes
