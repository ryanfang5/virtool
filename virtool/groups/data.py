from typing import List, Dict

from pymongo.errors import DuplicateKeyError

from virtool.data.errors import ResourceNotFoundError, ResourceConflictError
from virtool.groups.db import update_member_users, fetch_group_users
from virtool.mongo.core import DB
from virtool.mongo.utils import get_one_field
from virtool.users.utils import generate_base_permissions
from virtool.utils import base_processor
from virtool_core.models.group import GroupMinimal, Group


class GroupsData:
    def __init__(self, db):
        self._db: DB = db

    async def find(self) -> List[GroupMinimal]:
        """
        List all user groups.

        :returns: a list of all user groups

        """
        return [
            GroupMinimal(**base_processor(document))
            async for document in self._db.groups.find()
        ]

    async def get(self, group_id: str) -> Group:
        """
        Get a single group by its ID.

        :param group_id: the group's ID
        :return: the group
        """
        document = await self._db.groups.find_one(group_id)

        if document is None:
            raise ResourceNotFoundError()

        return Group(
            **base_processor(document),
            users=await fetch_group_users(self._db, group_id)
        )

    async def create(self, group_id: str) -> Group:
        """
        Create new group given a group ID.

        :param group_id: the ID for the new group
        :return: the group
        """
        try:
            document = await self._db.groups.insert_one(
                {
                    "_id": group_id,
                    "permissions": generate_base_permissions(),
                }
            )
        except DuplicateKeyError:
            raise ResourceConflictError("Group already exists")

        return Group(**base_processor(document), users=[])

    async def update(self, group_id: str, permissions_update: Dict[str, str]) -> Group:
        """
        Update the permissions for a group.

        :param group_id: the ID of the group to update
        :param permissions_update: the update to merge into the current permissions
        :return: the group
        """
        permissions = await get_one_field(
            self._db.groups, "permissions", {"_id": group_id}
        )

        if not permissions:
            raise ResourceNotFoundError

        async with self._db.create_session() as session:
            document = await self._db.groups.find_one_and_update(
                {"_id": group_id},
                {"$set": {"permissions": {**permissions, **permissions_update}}},
                session=session,
            )

            await update_member_users(self._db, group_id, session=session)

        return Group(
            **base_processor(document),
            users=await fetch_group_users(self._db, group_id)
        )

    async def delete(self, group_id: str):
        async with self._db.create_session() as session:
            delete_result = await self._db.groups.delete_one(
                {"_id": group_id}, session=session
            )

            if not delete_result.deleted_count:
                raise ResourceNotFoundError

            await update_member_users(self._db, group_id, remove=True, session=session)
