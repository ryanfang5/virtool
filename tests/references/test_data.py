import pytest
from aiohttp.test_utils import make_mocked_coro

from virtool.data.utils import get_data_from_app


# async def test_create_manifest(dbi, test_otu, app, data_layer):
#     await dbi.otus.insert_many(
#         [
#             test_otu,
#             dict(test_otu, _id="foo", version=5),
#             dict(test_otu, _id="baz", version=3, reference={"id": "123"}),
#             dict(test_otu, _id="bar", version=11),
#         ]
#     )
#     app["data"] = data_layer
#
#     assert await get_data_from_app(app).references.get_manifest("hxn167") == {
#         "6116cba1": 0,
#         "foo": 5,
#         "bar": 11,
#     }

