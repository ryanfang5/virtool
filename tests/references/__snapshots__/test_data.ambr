# name: TestEdit.test_control[uvloop--False]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': None,
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop--False].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': '',
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop--True]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': <class 'dict'> {
      'id': 'baz',
    },
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop--True].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': '',
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-None-False]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': None,
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-None-False].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': <class 'dict'> {
      'id': 'bar',
    },
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-None-True]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': <class 'dict'> {
      'id': 'baz',
    },
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-None-True].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': <class 'dict'> {
      'id': 'bar',
    },
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-baz-False]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': None,
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-baz-False].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': 'baz',
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-baz-True]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'id': 'foo',
    'internal_control': <class 'dict'> {
      'id': 'baz',
    },
    'latest_build': None,
    'name': 'Tester',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_control[uvloop-baz-True].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'description': 'This is a test reference.',
    'internal_control': 'baz',
    'name': 'Tester',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_reference_name[uvloop]
  <class 'dict'> {
    'contributors': <class 'list'> [
    ],
    'data_type': 'genome',
    'id': 'foo',
    'internal_control': None,
    'latest_build': None,
    'name': 'Bar',
    'otu_count': 0,
    'unbuilt_change_count': 0,
    'user': <class 'dict'> {
      'administrator': False,
      'handle': 'leeashley',
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'zclark',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_reference_name[uvloop].1
  <class 'dict'> {
    '_id': 'foo',
    'data_type': 'genome',
    'internal_control': <class 'dict'> {
      'id': 'bar',
    },
    'name': 'Bar',
    'user': <class 'dict'> {
      'id': 'bf1b993c',
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: TestEdit.test_reference_name[uvloop].2
  <class 'list'> [
    <class 'dict'> {
      '_id': 'baz',
      'reference': <class 'dict'> {
        'id': 'foo',
        'name': 'Bar',
      },
    },
    <class 'dict'> {
      '_id': 'boo',
      'reference': <class 'dict'> {
        'id': 'foo',
        'name': 'Bar',
      },
    },
  ]
---